Objects/perf_trampoline.c

```diff
@@ -0,0 +1,165 @@
+                        0);  // offset (not used here)
+    if (!memory) {
+        Py_FatalError("Failed to allocate new code arena");
+        return -1;
+    }
+    void *start = &_Py_trampoline_func_start;
+    void *end = &_Py_trampoline_func_end;
+    size_t code_size = end - start;
+
+    long n_copies = page_size / code_size;
+    for (int i = 0; i < n_copies; i++) {
+        memcpy(memory + i * code_size, start, code_size * sizeof(char));
+    }
+
+    mprotect(memory, page_size, PROT_READ | PROT_EXEC);
+
+    code_arena.start_addr = memory;
+    code_arena.current_addr = memory;
+    code_arena.size = page_size;
+    code_arena.size_left = page_size;
+    code_arena.code_size = code_size;
+    return 0;
+}
+
+static inline py_trampoline
+code_arena_new_code(code_arena_t *code_arena)
+{
+    py_trampoline trampoline = (py_trampoline)code_arena->current_addr;
+    code_arena->size_left -= code_arena->code_size;
+    code_arena->current_addr += code_arena->code_size;
+    return trampoline;
+}
+
+static inline py_trampoline
+compile_trampoline(void)
+{
+    if (code_arena.size_left <= code_arena.code_size) {
+        if (new_code_arena() < 0) {
+            return NULL;
+        }
+    }
+
+    assert(code_arena.size_left <= code_arena.size);
+    return code_arena_new_code(&code_arena);
+}
+
+static inline FILE *
+perf_map_open(pid_t pid)
+{
+    char filename[100];
+    snprintf(filename, sizeof(filename), "/tmp/perf-%d.map", pid);
+    FILE *res = fopen(filename, "a");
+    if (!res) {
+        _Py_FatalErrorFormat(__func__, "Couldn't open %s: errno(%d)", filename, errno);
+        return NULL;
+    }
+    return res;
+}
+
+static inline int
+perf_map_close(FILE *fp)
+{
+    if (fp) {
+        return fclose(fp);
+    }
+    return 0;
+}
+
+static void
+perf_map_write_entry(FILE *method_file, const void *code_addr,
+                     unsigned int code_size, const char *entry,
+                     const char *file)
+{
+    fprintf(method_file, "%lx %x py::%s:%s\n", (unsigned long)code_addr,
+            code_size, entry, file);
+}
+
+static PyObject *
+py_trampoline_evaluator(PyThreadState *ts, _PyInterpreterFrame *frame,
+                        int throw)
+{
+    PyCodeObject *co = frame->f_code;
+    py_trampoline f = NULL;
+    _PyCode_GetExtra((PyObject *)co, extra_code_index, (void **)&f);
+    if (f == NULL) {
+        if (extra_code_index == -1) {
+            extra_code_index = _PyEval_RequestCodeExtraIndex(NULL);
+        }
+        py_trampoline new_trampoline = compile_trampoline();
+        if (new_trampoline == NULL) {
+            return NULL;
+        }
+        FILE *pfile = perf_map_open(getpid());
+        if (pfile == NULL) {
+            return NULL;
+        }
+        perf_map_write_entry(pfile, new_trampoline, code_arena.code_size,
+                             PyUnicode_AsUTF8(co->co_qualname),
+                             PyUnicode_AsUTF8(co->co_filename));
+        perf_map_close(pfile);
```

Comment:
Is it necessary to open and close the file for each entry? Could we keep the file open, synchronize writes with a lock (GIL?) and just fflush?

Also do we have to keep the file around when the program exits? Test files are flooding my tmp dir.

Reply:
> Also do we have to keep the file around when the program exits? Test files are flooding my tmp dir.

Yes, otherwise `perf report` or other reporting commands won't work. The file is inspected on report time, not on collection time.

Reply:
> Is it necessary to open and close the file for each entry? Could we keep the file open, synchronize writes with a lock (GIL?) and just fflush?

So we you suggest that we don't close it? This operation is not on the hot path, just the first time a function is called so I think this is not really a problem. But if you want we can always do as you suggest

Reply:
How are nested functions handled? I cannot remember whether Python generates a new code object for every creating of an inner function or not.

Reply:
Same code object, different function object. This works nicely, even for list comprehensions and other forms of explicit or implicit scope.